File: src/lambda/document-processor/utils/s3-client.ts
/**
 * S3 Client Utilities for Document Processing
 *
 * Provides S3 operations for reading source documents and writing
 * converted images. Uses presigned URLs for secure access.
 *
 * Storage Path Pattern: {business_id}/{user_id}/{document_type}/{document_id}/{stage}/{filename}
 * Processing stages: raw, converted, processed
 */

import {
  S3Client,
  GetObjectCommand,
  PutObjectCommand,
  HeadObjectCommand,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import type { ConvertedImageInfo } from '../types';

// Storage path types
export type DocumentType = 'invoice' | 'receipt' | 'other';
export type ProcessingStage = 'raw' | 'converted' | 'processed';

// S3 client singleton (reused across warm invocations)
const s3Client = new S3Client({
  region: process.env.AWS_REGION || 'us-west-2',
});

const BUCKET_NAME = process.env.S3_BUCKET_NAME || 'finanseal-bucket';
const PRESIGNED_URL_EXPIRY = 3600; // 1 hour

/**
 * Error thrown when S3 operations fail
 */
export class S3OperationError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly key?: string
  ) {
    super(message);
    this.name = 'S3OperationError';
  }
}

/**
 * Read a document from S3 as a Buffer.
 *
 * @param key - S3 object key (e.g., 'invoices/doc-123.pdf')
 * @returns Document content as Buffer
 * @throws {S3OperationError} If read fails
 */
export async function readDocument(key: string): Promise<Buffer> {
  try {
    const command = new GetObjectCommand({
      Bucket: BUCKET_NAME,
      Key: key,
    });

    const response = await s3Client.send(command);

    if (!response.Body) {
      throw new S3OperationError(
        'Empty response body from S3',
        'S3_READ_ERROR',
        key
      );
    }

    // Convert readable stream to Buffer
    const chunks: Uint8Array[] = [];
    for await (const chunk of response.Body as AsyncIterable<Uint8Array>) {
      chunks.push(chunk);
    }

    return Buffer.concat(chunks);
  } catch (error) {
    if (error instanceof S3OperationError) throw error;

    const awsError = error as Error & { name?: string };
    if (awsError.name === 'NoSuchKey') {
      throw new S3OperationError(
        `Document not found in S3: ${key}`,
        'S3_NOT_FOUND',
        key
      );
    }

    throw new S3OperationError(
      `Failed to read document from S3: ${awsError.message}`,
      'S3_READ_ERROR',
      key
    );
  }
}

/**
 * Get document metadata without downloading content.
 *
 * @param key - S3 object key
 * @returns Object metadata including content type and size
 */
export async function getDocumentMetadata(key: string): Promise<{
  contentType: string;
  contentLength: number;
  lastModified?: Date;
}> {
  try {
    const command = new HeadObjectCommand({
      Bucket: BUCKET_NAME,
      Key: key,
    });

    const response = await s3Client.send(command);

    return {
      contentType: response.ContentType || 'application/octet-stream',
      contentLength: response.ContentLength || 0,
      lastModified: response.LastModified,
    };
  } catch (error) {
    const awsError = error as Error & { name?: string };
    if (awsError.name === 'NotFound' || awsError.name === 'NoSuchKey') {
      throw new S3OperationError(
        `Document not found in S3: ${key}`,
        'S3_NOT_FOUND',
        key
      );
    }

    throw new S3OperationError(
      `Failed to get document metadata: ${awsError.message}`,
      'S3_READ_ERROR',
      key
    );
  }
}

/**
 * Generate standardized storage path for converted images.
 *
 * Pattern: {business_id}/{user_id}/{document_type}/{document_id}/converted/{filename}
 *
 * @param config - Storage path configuration
 * @returns S3 key following the standardized pattern
 */
export function generateStoragePath(config: {
  businessId: string;
  userId: string;
  documentType: DocumentType;
  documentId: string;
  stage: ProcessingStage;
  filename: string;
}): string {
  const { businessId, userId, documentType, documentId, stage, filename } = config;
  return `${businessId}/${userId}/${documentType}/${documentId}/${stage}/${filename}`;
}

/**
 * Write a converted image to S3.
 *
 * @param documentId - Original document ID
 * @param pageNumber - Page number (1-indexed)
 * @param imageBuffer - PNG image data
 * @param metadata - Image dimensions
 * @param pathConfig - Optional storage path configuration for proper hierarchy
 * @returns S3 key for the uploaded image
 */
export async function writeConvertedImage(
  documentId: string,
  pageNumber: number,
  imageBuffer: Buffer,
  metadata: { width: number; height: number },
  pathConfig?: {
    businessId: string;
    userId: string;
    documentType: DocumentType;
  }
): Promise<string> {
  const filename = `page-${pageNumber.toString().padStart(3, '0')}.png`;

  // Use proper hierarchical path if config provided, otherwise fallback to legacy
  const key = pathConfig
    ? generateStoragePath({
        businessId: pathConfig.businessId,
        userId: pathConfig.userId,
        documentType: pathConfig.documentType,
        documentId,
        stage: 'converted',
        filename,
      })
    : `converted/${documentId}/${filename}`; // Legacy fallback

  try {
    const command = new PutObjectCommand({
      Bucket: BUCKET_NAME,
      Key: key,
      Body: imageBuffer,
      ContentType: 'image/png',
      Metadata: {
        documentId,
        pageNumber: pageNumber.toString(),
        width: metadata.width.toString(),
        height: metadata.height.toString(),
      },
    });

    await s3Client.send(command);
    return key;
  } catch (error) {
    const awsError = error as Error;
    throw new S3OperationError(
      `Failed to write converted image: ${awsError.message}`,
      'S3_WRITE_ERROR',
      key
    );
  }
}

/**
 * Write multiple converted images to S3.
 *
 * @param documentId - Original document ID
 * @param images - Array of image data with metadata
 * @param pathConfig - Optional storage path configuration for proper hierarchy
 * @returns Array of ConvertedImageInfo with S3 keys
 */
export async function writeConvertedImages(
  documentId: string,
  images: Array<{
    pageNumber: number;
    imageBuffer: Buffer;
    width: number;
    height: number;
  }>,
  pathConfig?: {
    businessId: string;
    userId: string;
    documentType: DocumentType;
  }
): Promise<ConvertedImageInfo[]> {
  const results: ConvertedImageInfo[] = [];

  // Process images in parallel (up to 5 concurrent)
  const batchSize = 5;
  for (let i = 0; i < images.length; i += batchSize) {
    const batch = images.slice(i, i + batchSize);
    const batchResults = await Promise.all(
      batch.map(async (img) => {
        const s3Key = await writeConvertedImage(
          documentId,
          img.pageNumber,
          img.imageBuffer,
          { width: img.width, height: img.height },
          pathConfig // Pass path config for proper hierarchy
        );

        return {
          pageNumber: img.pageNumber,
          s3Key,
          width: img.width,
          height: img.height,
          sizeBytes: img.imageBuffer.length,
        };
      })
    );
    results.push(...batchResults);
  }

  return results;
}

/**
 * Generate a presigned URL for reading a document.
 *
 * @param key - S3 object key
 * @param expiresIn - URL expiry in seconds (default 1 hour)
 * @returns Presigned URL for GET operation
 */
export async function getPresignedReadUrl(
  key: string,
  expiresIn: number = PRESIGNED_URL_EXPIRY
): Promise<string> {
  const command = new GetObjectCommand({
    Bucket: BUCKET_NAME,
    Key: key,
  });

  return getSignedUrl(s3Client, command, { expiresIn });
}

/**
 * Generate presigned URLs for multiple converted images.
 *
 * @param images - Array of ConvertedImageInfo
 * @returns Array of presigned URLs in same order
 */
export async function getPresignedImageUrls(
  images: ConvertedImageInfo[]
): Promise<string[]> {
  return Promise.all(
    images.map((img) => getPresignedReadUrl(img.s3Key))
  );
}

/**
 * Determine file type from content type or extension.
 *
 * @param key - S3 object key
 * @param contentType - Optional content type from metadata
 * @returns 'pdf' or 'image'
 */
export function determineFileType(
  key: string,
  contentType?: string
): 'pdf' | 'image' {
  if (contentType) {
    if (contentType === 'application/pdf') return 'pdf';
    if (contentType.startsWith('image/')) return 'image';
  }

  const extension = key.split('.').pop()?.toLowerCase();
  if (extension === 'pdf') return 'pdf';
  if (['jpg', 'jpeg', 'png', 'gif', 'webp', 'tiff'].includes(extension || '')) {
    return 'image';
  }

  // Default to image if unknown
  return 'image';
}

File: src/lambda/document-processor-python/utils/s3_client.py
"""
S3 Client for Python Lambda

Provides S3 operations for document storage including:
- Reading original documents
- Writing converted images
- Generating presigned URLs for image access
"""

import io
import os
from typing import List, Optional, Tuple
from dataclasses import dataclass

import boto3
from botocore.config import Config


@dataclass
class ConvertedImageInfo:
    """Information about a converted image."""
    page_number: int
    s3_key: str
    width: int
    height: int
    mime_type: str = "image/png"


class S3Client:
    """S3 client for document storage operations."""

    def __init__(
        self,
        bucket: str,
        region: str = "us-west-2",
        presign_expiration: int = 3600,
    ):
        """
        Initialize S3 client.

        Args:
            bucket: S3 bucket name
            region: AWS region
            presign_expiration: Presigned URL expiration in seconds
        """
        self.bucket = bucket
        self.region = region
        self.presign_expiration = presign_expiration

        # Configure client with retries
        config = Config(
            region_name=region,
            retries={"max_attempts": 3, "mode": "adaptive"},
        )
        self._client = boto3.client("s3", config=config)

    def read_document(self, s3_key: str) -> bytes:
        """
        Read document content from S3.

        Args:
            s3_key: S3 object key

        Returns:
            Document bytes

        Raises:
            S3Error: If read fails
        """
        try:
            response = self._client.get_object(Bucket=self.bucket, Key=s3_key)
            return response["Body"].read()
        except Exception as e:
            raise S3Error(f"Failed to read {s3_key}: {str(e)}")

    def write_image(
        self,
        s3_key: str,
        image_bytes: bytes,
        content_type: str = "image/png",
    ) -> str:
        """
        Write image to S3.

        Args:
            s3_key: S3 object key
            image_bytes: Image data
            content_type: MIME type

        Returns:
            S3 key of written object

        Raises:
            S3Error: If write fails
        """
        try:
            self._client.put_object(
                Bucket=self.bucket,
                Key=s3_key,
                Body=image_bytes,
                ContentType=content_type,
            )
            return s3_key
        except Exception as e:
            raise S3Error(f"Failed to write {s3_key}: {str(e)}")

    def get_presigned_url(self, s3_key: str, expiration: Optional[int] = None) -> str:
        """
        Generate presigned URL for reading an S3 object.

        Args:
            s3_key: S3 object key
            expiration: URL expiration in seconds (default: instance presign_expiration)

        Returns:
            Presigned URL

        Raises:
            S3Error: If URL generation fails
        """
        try:
            return self._client.generate_presigned_url(
                "get_object",
                Params={"Bucket": self.bucket, "Key": s3_key},
                ExpiresIn=expiration or self.presign_expiration,
            )
        except Exception as e:
            raise S3Error(f"Failed to generate presigned URL for {s3_key}: {str(e)}")

    def get_presigned_urls_for_images(
        self,
        images: List[ConvertedImageInfo],
    ) -> List[str]:
        """
        Generate presigned URLs for converted images.

        Args:
            images: List of converted image info

        Returns:
            List of presigned URLs
        """
        return [self.get_presigned_url(img.s3_key) for img in images]

    def build_storage_path(
        self,
        domain: str,
        business_id: str,
        user_id: str,
        document_id: str,
        stage: str,
        filename: str,
    ) -> str:
        """
        Build standardized S3 storage path (full S3 key).

        Pattern: {domain}/{business_id}/{user_id}/{document_id}/{stage}/{filename}

        This creates the complete S3 key including the domain prefix.
        Used by Lambda when writing converted images directly to S3.

        Aligned with TypeScript pattern:
        - TypeScript storage path (Convex): {bid}/{uid}/{docId}/{stage}/{filename}
        - TypeScript S3 key: {domain}/{bid}/{uid}/{docId}/{stage}/{filename}
        - Lambda S3 key: same as TypeScript S3 key

        Args:
            domain: 'invoices' or 'expense_claims'
            business_id: Business ID (Convex ID)
            user_id: User ID (Convex ID)
            document_id: Document ID (Convex ID)
            stage: Processing stage ('raw', 'converted', 'processed')
            filename: File name (e.g., 'page_1.png')

        Returns:
            Full S3 key path
        """
        return f"{domain}/{business_id}/{user_id}/{document_id}/{stage}/{filename}"

    def write_converted_images(
        self,
        images: List[Tuple[bytes, int, int]],
        document_id: str,
        domain: str,
        storage_path: str,
    ) -> List[ConvertedImageInfo]:
        """
        Write converted PDF page images to S3.

        Args:
            images: List of (image_bytes, width, height) tuples
            document_id: Document ID
            domain: 'invoices' or 'expense_claims'
            storage_path: Original storage path (for path construction)

        Returns:
            List of ConvertedImageInfo for each written image
        """
        results = []

        for page_num, (img_bytes, width, height) in enumerate(images, start=1):
            # Extract path components from original storage path
            # Format: {business_id}/{user_id}/{document_id}/raw/{filename}
            # Note: storage_path from Convex doesn't include domain prefix
            path_parts = storage_path.split("/")

            if len(path_parts) >= 3:
                business_id = path_parts[0]
                user_id = path_parts[1]
            else:
                business_id = "unknown"
                user_id = "unknown"

            # Build S3 key for converted image
            filename = f"page_{page_num}.png"
            s3_key = self.build_storage_path(
                domain=domain,
                business_id=business_id,
                user_id=user_id,
                document_id=document_id,
                stage="converted",
                filename=filename,
            )

            # Write to S3
            self.write_image(s3_key, img_bytes, "image/png")

            results.append(ConvertedImageInfo(
                page_number=page_num,
                s3_key=s3_key,
                width=width,
                height=height,
                mime_type="image/png",
            ))

        return results

    def get_full_s3_key(self, storage_path: str, domain: str) -> str:
        """
        Build full S3 key from storage path and domain.

        The storage_path from Convex typically doesn't include the domain prefix.
        This method prepends the domain to create the full S3 key.

        Args:
            storage_path: Path without domain prefix
            domain: 'invoices' or 'expense_claims'

        Returns:
            Full S3 key with domain prefix
        """
        return f"{domain}/{storage_path}"

    def object_exists(self, s3_key: str) -> bool:
        """
        Check if an S3 object exists.

        Args:
            s3_key: S3 object key

        Returns:
            True if object exists
        """
        try:
            self._client.head_object(Bucket=self.bucket, Key=s3_key)
            return True
        except:
            return False


class S3Error(Exception):
    """Error from S3 operations."""
    pass
